sympy/printing/codeprinter.py

```diff
@@ -69,6 +69,18 @@ class CodePrinter(StrPrinter):
     # may be supported
     # function_to_rewrite : (function_to_rewrite_to, iterable_with_other_functions_required)
     _rewriteable_functions = {
```

Comment:
I guess I didn't realize this is already a thing. So given that this is already here, we can keep it.

Reply:
Not sure about the mpmath printer. I do not think so as all(?) functions in SymPy already have mpmath-support(?).

The point was to support code generation for languages that didn't have support for some functions, but could be rewritten to have support. And to unify that so that one do not have to copy the same code to every printer to expand it. I *think* that it is probably only now that it has become possibly  problematic. Most (all?) earlier rewrites shouldn't have this problem.

These are the current automatic rewrites btw.
https://github.com/sympy/sympy/blob/0bc2587aef09ab150bd408b7cd9e08baf7b5fd9c/sympy/printing/codeprinter.py#L71-L94

And somehow, the problem is not really in the printing, but in the rewrites themselves. So if it is done automatically or a user have to do it to be able to print is not fundamentally different (from some perspective).

Finally, for rewriting to happen someone should add the rule to the list above. So the while it can happen that a new function is added and rather than adding support for latex and pprint, the rewrite support is added, it seems quite unlikely. Especially since this is not really a well known functionality.

Reply:
Let's just leave this as it is for now. I didn't realize this was already happening. I think this should only happen if the function can't be printed at all. I brought up the broader question of how to do rewrites in the printers here https://github.com/sympy/sympy/issues/22277#issuecomment-1058481071, although maybe it should be a separate issue.

Reply:
It looks like `mpmath` has a bunch of those functions e.g. factorial, erfc, li, ff, sinc, ...

Maybe the rewrites should be put into groups in the CodePrinter class so that subclasses can choose which groups to use e.g.:
```python
class CodePrinter:
    _basic_rewrites = {
        'sinc': 'sin',
        ...
    }
    _reciprocal_trig_rewrites = {
        'sec': 'cos',
        ...
    }

class NumPyPrinter(CodePrinter):
    _rewrites = {
        **CodePrinter._basic_rewrites,
        **CodePrinter._reciprocal_trig_rewrites,
    }
```

Reply:
> I think this should only happen if the function can't be printed at all.

This is exactly how it is. These are used as a final resort:
https://github.com/sympy/sympy/blob/cd738be3bdde5f8a0e53bf71b4420146acff392a/sympy/printing/codeprinter.py#L433-L437

Reply:
Honestly these class level variables that are so commonly used in the printer classes are not great for subclassing. You either get all or nothing. And furthermore, since they are mutable containers, you can get a subclass modifying them and changing all other subclasses:

```py
>>> class A:
...     x = {
...     'a': 1,
...     }
>>> class B(A):
...     def __init__(self):
...         self.x['b'] = 2
>>> A().x
{'a': 1}
>>> B()
<__main__.B object at 0x10b3fff70>
>>> A().x
{'a': 1, 'b': 2}
```

For this specific case, though, as long as the rewrites are ignored when the printer actually knows how to handle a given function, then it shouldn't be a problem to just have one big list.

Reply:
One can consider making it immutable (`collections.namedtuple()` may work from a quick googling). AFAIK there is no code (added by me) that modifies it.